### ISSUE TITLE
Term.branch_head no longer triggering a rebuild

### ISSUE THREAD

**Issue by avsm** [docker/datakit#413]
Term.branch_head as used in https://github.com/avsm/mirage-ci/blob/master/src-bin/mirageCI.ml#L87 is no longer triggering rebuilds on a push.  I verified these steps with @samoht:

- start a ci.mirage.io instance
- wait for it to finish building
- push to mirage/opam-repository
- datakit successfully updates its head in the local CI checkout (so webhooks are working)
- no rebuild happens in the UI
- restart the CI instance
- rebuild begins with the correct mirage/opam-repository head

So it looks like Term.branch_head may have regressed for the case of additional repositories being watched.

**Comment by talex5** [comment]
Although the code for interacting with GitHub events changed recently, I don't see anything obvious that would cause this, and `branch_head` is tested by the unit-tests at https://github.com/docker/datakit/blob/070c6a21e3ec315dd1315236f141d3b8b4ff7c80/ci/tests/test_ci.ml#L363.

Are the logs pushed somewhere? Would be good to be able to reproduce it by cloning the logs repo at some point, running the CI, and cherry-picking the update that caused the problem.

**Comment by avsm** [comment]
I'll set it up to push again; confirmed that the problem happened with a merge to mirage/mirage-dev just now and:

```
ci_1           | 2016-12-14 14:12.01 INF [datakit-ci] 200 - GET /fonts/glyphicons-halflings-regular.woff2
bridge_1       | datakit-github: [INFO] received webhook event for mirage/mirage-dev: pull-request
bridge_1       | datakit-github: [INFO] Processing new entry -- active repos: mirage/mirage
bridge_1       |                        mirage/mirage-dev mirage/opam-repository
bridge_1       | datakit-github: [INFO] received webhook event for mirage/mirage-dev: push
bridge_1       | datakit-github: [INFO] API.status {mirage/mirage-dev 77541f52ade5b554768e558196395db720806c52}
bridge_1       | datakit-github: [INFO] Processing new entry -- active repos: mirage/mirage
bridge_1       |                        mirage/mirage-dev mirage/opam-repository
bridge_1       | datakit-github: [INFO] Change detected in github-metadata
bridge_1       | datakit-github: [INFO] Processing new entry -- active repos: mirage/mirage
bridge_1       |                        mirage/mirage-dev mirage/opam-repository
bridge_1       | datakit-github: [INFO] received webhook event for mirage/mirage-dev: status
bridge_1       | datakit-github: [INFO] Processing new entry -- active repos: mirage/mirage
bridge_1       |                        mirage/mirage-dev mirage/opam-repository
bridge_1       | datakit-github: [INFO] Change detected in github-metadata
bridge_1       | datakit-github: [INFO] Processing new entry -- active repos: mirage/mirage
bridge_1       |                        mirage/mirage-dev mirage/opam-repository
```

**Comment by avsm** [comment]
Its now rebuilding on startup again and autopushing to https://github.com/mirage/mirage-ci.logs

**Comment by talex5** [comment]
Might be worth setting the `datakit-ci` source to DEBUG level (you can use `-vv`on the command-line, but that also enables 9p debug which might be a bit verbose).

**Comment by talex5** [comment]
OK, I can reproduce this locally. Investigating...

**Comment by talex5** [comment]
Here's the problem: https://github.com/avsm/mirage-ci/blob/master/src/opam_build.ml#L35

You need to compare all the fields in `key` to check if two keys are equal (in particular, `remote_git_rev` isn't checked).

I usually use pattern matching on one of the arguments so I can't forget anything.

**Comment by avsm** [comment]
ahhh! Thanks @talex5!

**Comment by avsm** [comment]
fixed in avsm/mirage-ci@8d17491a2d6e23f833de489e867df133a3218332
